openshift/machine-config-operator
# MCO-1326: Decouple BuildController and ImageBuildRequest

**- What I did**

This is the first part of my unwedge BuildController work. This refactors and reorganizes a lot of the difficult-to-reason-about and difficult-to-test code paths in BuildController into separate modules so that BuildController itself can be more easily extended and worked on. Care was taken to ensure that all BuildRequest (formerly ImageBuildRequest) code is well-scoped, well-tested, and only exposes a very limited interface to BuildController and other places where it might be used.

The code is broken up thusly:
- There is a new `constants` subpackage within `pkg/controller/build` which contains all of the widely-used and well-known constants such as annotation keys, label keys, etc. In addition,  this package has a file called `selectors.go` that can construct a `LabelSelector` with the keys which can be useful for performing label queries.
- There is anew `buildrequest` subpackage within `pkg/controller/build` which contains all of the code and tests needed to generate a BuildRequest (formerly called `ImageBuildRequest`). Of particular note is the `Preparer` object. This object is capable of creating all of the ephemeral build objects as well as being able to destroy them. By doing this, BuildController can be significantly trimmed down.
- The only major functionality change is that `Preparer` will create copies of the configured image pull secrets for each image build. It will remove those copies upon completion of the build.
- It is also worth mentioning that all code under `buildrequest` expects a `context.Context` to be passed into it. For now, BuildController will just use a `context.TODO()`, but a future update will allow it to create a child context attached to the parent context.

**- How to verify it**

As mentioned above, BuildController functionality should be the same. Therefore, the preexisting automated test suites should be able to verify the change.

**- Description for the changelog**
Decouple BuildController and ImageBuildRequest
